Ultimate conditional syntax

chengluyu opened the pull request.

LPTK committed:
Add variable symbols inserted by Typer

chengluyu committed:
Wire the desugaring logic with the typer

chengluyu committed:
Make some tests work

chengluyu committed:
Refactor DiffTests so code generation is after type checking

There is some annoying blank lines. I don't know why.

LPTK committed:
Add some currently failing tests

chengluyu committed:
Implement code generation for the `new` syntax

chengluyu committed:
Revert removing a DiffTests regression

chengluyu committed:
Report error if field `els` is missing in `If`

chengluyu committed:
Pretty-print debug outputs

chengluyu committed:
Fix some test cases

TBH, my tree construction logic is quite weird.

chengluyu committed:
Show the wrong syntax tree by `:p`

chengluyu committed:
Migrate test cases from the old branch

chengluyu committed:
Wrap some test cases in functions

chengluyu committed:
Handle more UCS cases

- Make use of partial terms in `desugarIfBody`.
- Make use of partial patterns in `desugarMatchBranch`.
- Handle more `IfBody`s. So, it's more than SUCS.
- "And"-separation goes deeper.
- Extract minor desugaring routines.
- Some other minor changes...

chengluyu committed:
Rectify a false positive

chengluyu committed:
Handle wildcard cases and dedup let bindings

chengluyu committed:
Fix cases in `DirectLines.mls`

chengluyu committed:
Review test cases in `Humiliation.mls`

chengluyu committed:
Implement basic exhaustiveness check

chengluyu committed:
Suppor tuple pattern syntax and improve defualt case propagation

chengluyu committed:
Handle `IfOpsApp` at the pattern position

chengluyu committed:
Review test cases

chengluyu committed:
Fix my stupid double negation phrasing

chengluyu committed:
Check the test case mentioned in Telegram chat

chengluyu committed:
Add missing test file changes

chengluyu committed:
Embed UCS related code to a layer in the cake pattern

chengluyu committed:
Get rid of `Typer#Ctx`

chengluyu committed:
Adjust methods due to the previous refactor

chengluyu committed:
Reduce boilerplate by implementing an ADT for `PartialTerm`

chengluyu committed:
Organize branches in `desugarIf`

chengluyu committed:
Implement interleaved let bindings

- Localize scrutinee if it is not a `Var`
- Support basic interleaved let bindings
- Fix test cases in InterleavedLet.mls
- Keep ordering when accumulating bindings
- Support interleaved let bindings in default branches
- Simplify `MutCaseOf.Match` by separating the wildcard
- Support reordering conditions

chengluyu committed:
Localize scrutinee if it is not a `Var`

chengluyu committed:
Merge branch 'ucs-translation' of github.com:hkust-taco/mlscript into ucs-translation

chengluyu committed:
Gather parser failure examples

chengluyu committed:
Prepare for improving error messages

chengluyu committed:
Improve error messages but it's still far from perfect

chengluyu committed:
Inspect test files and mark problematic tests

LPTK commented on shared/src/test/diff/nu/SimpleUCS.mls:
```diff
@@ -223,20 +252,25 @@ fun f(u, v, w) =
+//│ ║  l.250: 	    Some(x) and x is
+//│ ║         	                ^
+//│ ╟── from field selection:
+//│ ║  l.249: 	  if foo(u, v, w) is
```
Whoops, this is not a field selection!

chengluyu replied:
This is done by adding a `sugaredTerm` to `TermImpl`. It's not perfect either.

LPTK replied:
Seems good enough for now!

LPTK commented on shared/src/test/diff/nu/InterleavedLet.mls:
```diff
@@ -235,29 +179,163 @@ fun tr(x) =
+//│ Cons: ('head, 'tail,) -> (Cons with {head: 'head, tail: 'tail})
+//│     = [Function: Cons1]
+
+// FIXME: Take a look...
```
It should be `Nil then Pair(Nil(), Nil())`.

chengluyu replied:
This one is fixed.

LPTK commented on shared/src/test/diff/nu/SimpleUCS.mls:
```diff
@@ -129,17 +116,20 @@ fun f(x, y) =
 //│ Code generation encountered an error:
 //│   if expression has not been desugared
 
+// FIXME
```
Yeah, this should be an exhaustiveness error (or at the very least a "useless case" warning for the `None() then xv * 2` part.

chengluyu replied:
This is fixed.

LPTK commented on shared/src/test/diff/nu/SimpleUCS.mls:
```diff
@@ -234,49 +208,48 @@ fun f(x, a, b) =
+fun f(x, y, b) =
   if x is
     Some(xv) and y
       is Some(yv) then "bruh"
```
Curious to see what happens if you make it
```suggestion
      is Some(yv) then yv
```
(please add this as an additional test)

chengluyu replied:
It's added.

LPTK commented on shared/src/test/diff/nu/SplitOps.mls:
```diff
@@ -79,37 +109,18 @@ fun f(x) =
+//│       = [Function: sumOpt]
 
-fun f(x, y) =
+// FIXME
```
This is legit: `int & B` is `nothing`. I guess it would be nice if the UCS could warn that something's off. I think the UCS should always complain when there is an `is` test on some scrutinee with only one associated pattern like here.

chengluyu replied:
This is implemented but does not work on some cases. I made a note of this and will improve it later.

LPTK replied:
Ok, please add the missing case description to https://github.com/hkust-taco/mlscript/issues/151

LPTK commented:
Even though they're not perfect yet, the errors are pretty cool!

chengluyu committed:
Add an additional test suggested in the PR

chengluyu committed:
Fix the `mapPartition` test and add some related tests

chengluyu committed:
Implement a basic solution to type errors at desugared terms

chengluyu committed:
Split `UltimateConditions` into smaller files

chengluyu committed:
Tracking locations everywhere!

This makes the code even more messy.

chengluyu committed:
Show `IfBody` in the `inspect` method

chengluyu committed:
Move UCS tests to a new folder

chengluyu committed:
Fix the exhaustiveness checker

chengluyu committed:
Fix that UCS expressions return nothing

chengluyu committed:
Warn of pattern matching with only one case

LPTK commented:
Beautiful!

chengluyu committed:
Support nested tuple patterns produced by the Ocaml parser

LPTK committed:
Add some tests from previous meeting

chengluyu committed:
Clean up tests

chengluyu committed:
Merge branch 'ucs-translation' of github.com:chengluyu/mlscript into ucs-translation

chengluyu committed:
Add locations to duplicated else branch warnings

chengluyu committed:
Merge branch 'mlscript' into ucs-translation

chengluyu committed:
Commit changed files changed by the GitHub merge

chengluyu committed:
Add more changed tests

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -715,8 +721,29 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool)
+          typeTerm(trm)
+        } catch {
+          case e: DesugaringException => e.report(this)
+          case e: Throwable => throw e
```
This line is useless.

LPTK commented on shared/src/test/diff/nu/ParserFailures.mls:
```diff
@@ -0,0 +1,16 @@
+//   if x
+//     is A() then "A"
+//     let y = 0
+//     is B() then "B"
```
The correct thing to do is NOT to comment those tests, but to add a `// FIXME` on top, which will silence the test failure.

LPTK replied:
You forgot to uncomment the tests 🤦‍♂️

chengluyu replied:
They will raise `???`. Can DiffTests catch `NotImplementedError`?

LPTK replied:
What part of "`// FIXME` will silence the test failure" don't you understand? 😜

chengluyu replied:
Oh. That's pretty cool.

LPTK commented on shared/src/main/scala/mlscript/helpers.scala:
```diff
@@ -7,6 +7,7 @@ import scala.collection.immutable.SortedSet
 import math.Ordered.orderingToOrdered
 
 import mlscript.utils._, shorthands._
+import scala.collection.immutable
```
This seems unused.

LPTK commented on shared/src/main/scala/mlscript/ucs/Clause.scala:
```diff
@@ -0,0 +1,89 @@
+
+  final case class BooleanTest(test: Term) extends Clause
+
+  // Make it a class, and then include methods in `Conjunction`?
```
Please either do this or remove the comment.

LPTK commented on shared/src/main/scala/mlscript/ucs/Clause.scala:
```diff
@@ -0,0 +1,89 @@
+    * @param interleavedLets the interleaved let buffer
+    * @return idential to `conditions`
+    */
+  def withBindings
+    (conditions: Conjunction)
+    (implicit interleavedLets: Buffer[(Bool, Var, Term)])
+  : Conjunction = {
```
```suggestion
  def withBindings(conditions: Conjunction)(implicit interleavedLets: Buffer[(Bool, Var, Term)])
        : Conjunction = {
```

LPTK commented on shared/src/main/scala/mlscript/ucs/Conjunction.scala:
```diff
@@ -0,0 +1,54 @@
+          } else {
+            rec(past :+ head, tail)
+          }
+        // TODO: Support tuples after merging `MatchClass` and `MatchTuple`.
```
Aren't tuples supported? If not, add to https://github.com/hkust-taco/mlscript/issues/151

chengluyu replied:
I didn't know how to check tuples when I wrote this. I though I could merge `MatchClass` with `MatchTuple`. As what we discussed, we don't need to merge them anymore. So, this is no longer in need.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+
+  /**
+    * Create a `Scrutinee`. If the `term` is a simple expression (e.g. `Var` or
+    * `Lit`), we do not create a local alias. Otherwise, we create a local alias
```
`Lit` doesn't actually seem to be handled.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+        case _: Var => Scrutinee(N, term)
+        case _ =>
+          Scrutinee(
+            S(localizedScrutineeMap.getOrElseUpdate(term, {
```
This seems like a pretty bad idea. What of two subterms compare equal for some reason? You'll assign them the same variable! Please find a cleaner and **robust** way to handle this.

chengluyu replied:
I'm adding `Loc` to the key of the map. Now it's `MutMap.empty[(Opt[Loc], Term), Var]`.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+    * do not contain interleaved let bindings.
+    */
+  private def destructPattern
+      (scrutinee: Term, pattern: Term, isMultiLineMatch: Bool = false)
```
Remove unnecessary default value. Defaults are error-prone so they should be reserved for easy-to-use APIs, not for core logic.

chengluyu replied:
Removed.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+              })
+            }
+        }
+        // if (branches.length === 1 && scrutinee.isMultiLineMatch && default.isEmpty) {
+        //   import Message.MessageContext
+        //   raise(WarningReport({
+        //     msg"This scrutinee has only one case." -> scrutinee.matchRootLoc
+        //   } :: Nil))
+        // }
```
What is this commented code doing here?

chengluyu replied:
It was for reporting pattern matching with only one case. The commented code here caused some false positive. I will remove it.

LPTK replied:
I added this to https://github.com/hkust-taco/mlscript/issues/151

chengluyu replied:
Thanks!

LPTK commented on shared/src/main/scala/mlscript/ucs/Scrutinee.scala:
```diff
@@ -0,0 +1,82 @@
+final case class Scrutinee(local: Opt[Var], term: Term) {
+  def reference: Var = local.getOrElse(term match {
+    case v: Var => v
+    case _      => ???
```
Is this a TODO? Either handle properly or add to https://github.com/hkust-taco/mlscript/issues/151

LPTK commented on shared/src/main/scala/mlscript/ucs/Scrutinee.scala:
```diff
@@ -0,0 +1,82 @@
+import mlscript.{Loc, Term, Var}
+import mlscript.utils.shorthands._
+
+// The point is to remember where does the scrutinee come from.
```
```suggestion
// The point is to remember where the scrutinee comes from.
```

LPTK commented on shared/src/main/scala/mlscript/ucs/Scrutinee.scala:
```diff
@@ -0,0 +1,82 @@
+    * Why we need a dedicated class for this?
+    * Because the scrutinee might comes from different structures.
+    * 1. Terms. E.g. `if x is A and y is B then ...`
+    * 2. `IfOpApp`. In this case, the scrutinee is indicate by the `IfOpApp`.
```
```suggestion
    * 2. `IfOpApp`. In this case, the scrutinee is indicated by the `IfOpApp`.
```

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+            term,
+          )
+      }
+      res.isMultiLineMatch = isMultiLineMatch
+      res.matchRootLoc = matchRootLoc
```
Do not use completely unnecessary mutation.

chengluyu replied:
`isMultiLineMatch` is removed.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+    * @param matchRootLoc the location of the root of the pattern matching
+    * @param fragments fragment term that used to construct the given pattern.
+    *   It is used to tracking locations.
+    * @param isMultiLineMatch whether the scrutinee is in multi-line pattern match.
```
I don't understand what this flag achieves. It doesn't seem to be really used anywhere.

chengluyu replied:
Removed.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+          case S(_) => 
+            val clause = Clause.MatchClass(makeScrutinee(scrutinee, isMultiLineMatch), classNameVar, Nil)
+            println(s"Build a Clause.MatchClass from $scrutinee where pattern is $classNameVar")
+            clause.locations = collectLocations(scrutinee)
```
Avoid unnecessary mutation. Why isn't this a parameter to `Clause`?

chengluyu replied:
Moved.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,605 @@
+        checkExhaustive(whenFalse, S(t))
+      case Match(scrutinee, branches, default) =>
+        scrutineePatternMap.get(getScurtineeKey(scrutinee)) match {
+          // Should I call `die` here?
```
`lastWords(s"...")` would be appropriate.

LPTK reviewed (changes requested):
This is great! But it still needs a little more polish.

It seems much of the logic hasn't been thoroughly cleaned up yet. Please do so. Try to remove unnecessary/gratuitious uses of mutable fields and variables.

I think too much info is currently printed in debug mode. Please add a `:ucsd` debugging flag to request more info. With only `:d`, I think we should basically only see two messages:

 * "The mutable CaseOf tree"

 * "Desugared term: ..."

LPTK committed:
Minor

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/Scrutinee.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/Scrutinee.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/Clause.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Handle `Lit` in `makeScrutinee`

chengluyu committed:
Remove useless re-throw

chengluyu committed:
Remove an useless import

chengluyu committed:
Mark two test cases as FIXME

chengluyu committed:
Call `lastWords` in some unreachable cases

chengluyu committed:
Remove an unused commented code

chengluyu committed:
Remove an outdated to-do

chengluyu committed:
Make `Conjunction` a class

chengluyu committed:
Identify scrutinees not only by terms but also by locations

chengluyu committed:
Move `Scrutinee.matchRootLoc` to the second parameter list

chengluyu committed:
Move `Clause.locations` to the second parameter list

chengluyu committed:
Hide minor UCS desugaring debug messages

LPTK commented on shared/src/test/diff/ucs/NestedPattern.mls:
```diff
@@ -0,0 +1,135 @@
+//│ f: error -> error
+
+:e
+:d
```
BTW we shouldn't commit debugging info.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -60,7 +60,7 @@ class Desugarer extends TypeDefs { self: Typer =>
   }
 
-  private val localizedScrutineeMap = MutMap.empty[Term, Var]
+  private val localizedScrutineeMap = MutMap.empty[(Opt[Loc], Term), Var]
```
I still hate this. What if a scrutinee doesn't have a location for some reason (eg. maybe it's synthetic)? The robustness of your code SHOULD NOT depend on locations being present in the trees!!

chengluyu replied:
Sure. Let me think what I can do.

LPTK replied:
If you really can't think of anything else, you can always try a `java.util.IdentityHashMap` https://docs.oracle.com/javase/7/docs/api/java/util/IdentityHashMap.html

chengluyu replied:
Actually `IdentityHashMap` doesn’t work in my case.

chengluyu replied:
The goal is to localize non-`Var` scrutinees by creating let bindings so that we can reduce unnecessary computations. For example, we'd like to translate
```
if foo(x, y, z)
  < 0 then left
  > 0 then right
  = _ then this
```
to
```
let t = foo(x, y, z) in
case t < 0 of
  true -> left
  false -> case t > 0 of
    true -> right
    false -> this
```
rather than
```
case foo(x, y, z) < 0 of
  true -> left
  false -> case foo(x, y, z) > 0 of
    true -> right
    false -> this
```

LPTK replied:
Why wouldn't `IdentityHashMap` work? I believe it does precisely what you want.

LPTK replied:
Actually I'm not really clear on why you need a map at all, but if you really do `IdentityHashMap` should work as we're talking about a specific non-duplicated term here. But my feeling is that you're doing something in a suboptimal way.

chengluyu replied:
This is good. I deleted two comments because there was a bug preventing it from working. Now it works.

chengluyu committed:
Remove debug output from UCS test files

chengluyu committed:
Remove dead code, add comments and die for unreachable cases

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -375,7 +375,8 @@ class Desugarer extends TypeDefs { self: Typer =>
         // This case is rare. Let's put it aside.
-        case L(IfLet(_, _, _, _)) => ???
+        case L(IfLet(_, _, _, _)) =>
+          lastWords("please add this rare case to test files")
```
Better leave `???` here. `lastWords` and `die` are for things that _cannot_ happen unless there's a bug in the compiler or in your reasoning.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -424,7 +425,8 @@ class Desugarer extends TypeDefs { self: Typer =>
           desugarIfBody(body, expr.addTermOp(lhs, op), acc)
         // This case is rare. Let's put it aside.
-        case IfLet(isRec, name, rhs, body) => ???
+        case IfLet(isRec, name, rhs, body) =>
```
Same.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -495,10 +497,11 @@ class Desugarer extends TypeDefs { self: Typer =>
               throw DesugaringException.Single(msg"Missing the otherwise case of test ${test.toString}", test.toLoc)
             } else {
-              ???
+              lastWords("`MissingCase` are not supposed to be the true branch of `IfThenElse`")
```
This one is a good use. (You can resolve this comment.)

chengluyu committed:
Revert two misuse of `lastWords`

chengluyu committed:
I just knew how to use `// FIXME` in DiffTests

chengluyu committed:
Use `IdentityHashMap` as a workaround

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -77,16 +78,18 @@ class Desugarer extends TypeDefs { self: Typer =>
+          val localName = if (localizedScrutineeMap.containsKey(term)) {
+            localizedScrutineeMap.get(term)
+          } else {
+            val v = Var(freshName).withLoc(term.toLoc)
```
Shouldn't this use `desugaredFrom`? In fact, maybe wee should _always_ use `desugaredFrom` instead of `withLocOf(term)` and we should make the latter also update the loc. (`withLocOf(term)` is currently like `withLocOf(term.toLoc)`).

chengluyu replied:
I didn't get "we should make the latter also update the loc". Doesn't `withLocOf(term)` update the loc?

LPTK replied:
Oh yeah I meant that we should make the _former_ update the loc, which is what you did.

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -375,8 +375,7 @@ class Desugarer extends TypeDefs { self: Typer =>
         // This case is rare. Let's put it aside.
-        case L(IfLet(_, _, _, _)) =>
-          lastWords("please add this rare case to test files")
+        case L(IfLet(_, _, _, _)) => ???
```
You know it's a pity that `???` does not have more info attached to it. You could add a method `def TODO(msg: Str) = throw new NotImplementedError(msg)` in `utils/package.scala`, and make this:
```diff
- case L(IfLet(_, _, _, _)) => ???
+ case L(IfLet(_, _, _, _)) => TODO("please add this rare case to test files")
```

chengluyu replied:
This is really cool. I added them.

chengluyu committed:
Add `TODO` and use it at rare cases

chengluyu committed:
Prefer `desugaredFrom` than `withLoc`

LPTK commented on shared/src/main/scala/mlscript/helpers.scala:
```diff
@@ -336,7 +336,7 @@ trait TermImpl extends StatementImpl { self: Term =>
   def desugaredFrom(term: Term): this.type = {
     sugaredTerm = S(term)
-    this
+    withLoc(term.toLoc)
```
```suggestion
    withLocOf(term)
```

LPTK commented on shared/src/main/scala/mlscript/ucs/Conjunction.scala:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  def separate(expectedScrutinee: Scrutinee): Opt[(MatchClass, Conjunction)] = {
+    def rec(past: Ls[Clause], upcoming: Ls[Clause])
+        : Opt[(Ls[Clause], MatchClass, Ls[Clause])] = {
```
```suggestion
    def rec(past: Ls[Clause], upcoming: Ls[Clause]): Opt[(Ls[Clause], MatchClass, Ls[Clause])] = {
```

LPTK commented on shared/src/main/scala/mlscript/ucs/Conjunction.scala:
```diff
@@ -0,0 +1,86 @@
+
+object Conjunction {
+  def empty: Conjunction = Conjunction(Nil, Nil)
+}
```
```suggestion
}

```

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,582 @@
+
+  import Clause.{MatchClass, MatchTuple, BooleanTest}
+
+  type FieldAliasMap = MutMap[Term, MutMap[Str, Var]]
```
This should probably be `MutMap[Var, MutMap[Str, Var]]` or `MutMap[SimpleTerm, MutMap[Str, Var]]`?

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -0,0 +1,582 @@
+    */
+  type ExhaustivenessMap = Map[Str \/ Int, Map[Var, MutCase]]
+
+  type MutExhaustivenessMap = MutMap[Str \/ Int, MutMap[Var, MutCase]]
```
Unused.

LPTK commented on shared/src/main/scala/mlscript/ucs/DesugaringException.scala:
```diff
@@ -0,0 +1,21 @@
+      typer.err(messages)
+    }
+  }
+}
```
```suggestion
}

```

LPTK commented on shared/src/main/scala/mlscript/ucs/DesugaringException.scala:
```diff
@@ -0,0 +1,21 @@
+    override def report(typer: Typer)(implicit raise: Diagnostic => Unit): typer.SimpleType = {
+      typer.err(message, location)
+    }
+  }
```
This seems useless. Just have one.

LPTK commented on shared/src/main/scala/mlscript/ucs/MutCaseOf.scala:
```diff
@@ -0,0 +1,370 @@
+
+  // A short-hand for pattern matchings with only true and false branches.
+  final case class IfThenElse(condition: Term, var whenTrue: MutCaseOf, var whenFalse: MutCaseOf) extends MutCaseOf {
+    override def describe: Str =
```
Remove all the unnecessary `override`s.

LPTK commented on shared/src/main/scala/mlscript/ucs/PartialTerm.scala:
```diff
@@ -0,0 +1,58 @@
+
+  final case class Half(lhs: Term, op: Var, override val fragments: Ls[Term]) extends PartialTerm(fragments) {
+    override def addTerm(rhs: Term): Total = op match {
+      // TODO: Why are they look the same? How did the code here evolve?
```
Please figure this out. Why *are* they look the same?

LPTK commented on shared/src/main/scala/mlscript/ucs/PartialTerm.scala:
```diff
@@ -0,0 +1,58 @@
+    override def addOp(op: Var): Half =
+      throw new PartialTermError(this, s"expect a term but operator $op was given")
+  }
+}
```
```suggestion
}

```

LPTK commented on shared/src/main/scala/mlscript/ucs/Scrutinee.scala:
```diff
@@ -0,0 +1,20 @@
+
+// The point is to remember where the scrutinee comes from.
+// Is it from nested patterns? Or is it from a `IfBody`?
+final case class Scrutinee(local: Opt[Var], term: Term)(implicit val matchRootLoc: Opt[Loc]) {
```
Remove unnecessary implicits unless they really reduce duplication.

LPTK commented on shared/src/test/diff/basics/Errors.fun:
```diff
@@ -126,6 +135,23 @@ log / false + 1
+
+
+
+
```
Is all this space necessary?

LPTK replied:
Please close when/if resolved.

LPTK commented on shared/src/test/diff/codegen/Terms.mls:
```diff
@@ -341,39 +342,14 @@ case A { a = 0 } of
 
+// FIXME
 // With
 {} with {}
-//│ res: anything
-//│    = {}
```
Is this change on purpose?

LPTK replied:
Please close when/if resolved.

LPTK reviewed (changes requested):
LGTM except that the test files are borked. Please proofread your diff!

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/Conjunction.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/Conjunction.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/PartialTerm.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Update shared/src/main/scala/mlscript/ucs/DesugaringException.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

chengluyu committed:
Remove unused `MutExhaustivenessMap`

chengluyu committed:
Simplify `DesugaringException`

chengluyu committed:
Remove duplicated code in `PartialTerm`

chengluyu committed:
Remove unnecessary `override` specifiers

chengluyu committed:
Remove unnecessary `override`

chengluyu committed:
Remove unnecessary implicits

chengluyu committed:
Update shared/src/main/scala/mlscript/helpers.scala

Co-authored-by: Lionel Parreaux <lionel.parreaux@gmail.com>

LPTK commented on shared/src/main/scala/mlscript/ucs/Desugarer.scala:
```diff
@@ -80,7 +80,7 @@ class Desugarer extends TypeDefs { self: Typer =>
   def makeScrutinee(term: Term)(implicit matchRootLoc: Opt[Loc]): Scrutinee =
```
There's still an implicit here... is it useful?

chengluyu committed:
Change the key of `FieldAliasMap` to `SimpleTerm`

Also, delete implicit parameters `matchRootLoc` since it's
included in the scrutinee.

chengluyu committed:
Check if name exists in `freshName`

chengluyu committed:
Add a failed case about split scrutinees

chengluyu committed:
Fix the bug shown by the previous commit

LPTK committed:
Add a couple of new tests

LPTK committed:
Fix parser precedences and generalize Term desugaring scheme

chengluyu committed:
It seems that blank lines are created by a solved bug

chengluyu committed:
Check out mlscript/*.mls from the main branch

chengluyu committed:
Append more test files

chengluyu committed:
Revert "Append more test files"

This reverts commit cb9c234b40ad404506d89b5eec12f9a4efe90960.

chengluyu committed:
Revert "Check out mlscript/*.mls from the main branch"

This reverts commit 2a7713c5ca59d59458edce093359024478ed1fad.

chengluyu committed:
Revert "It seems that blank lines are created by a solved bug"

This reverts commit ec8e7ccf7d2f090419f064a7417e2c72e689be35.

chengluyu committed:
Checkout files at basics, codegen and gen from the fork origin

These files are not changed in this PR surely.

chengluyu committed:
Checkout files at mlscript and typegen from the fork origin

chengluyu committed:
Remove blank lines from files at nu

chengluyu committed:
Prevent too much changes in PR #150

chengluyu committed:
Keep the original order between types, diagnosis and results

chengluyu committed:
Clean up DiffTests

LPTK commented:
Please make one last proofreading of your diff and then request the review again.

LPTK committed:
Minor UCS PR polishing

LPTK reviewed (approved):
Ok, it finally LGTM! 🥳

LPTK commented:
This time I'll create a merge commit because there's a lot of history in this branch.
